src/error.rs

```diff
@@ -1,141 +1,366 @@
+	MissingClosingBrace(MissingClosingBrace),
+
+	/// The input string contains a placeholder for a variable that is not in the variable map.
+	NoSuchVariable(NoSuchVariable),
 }
 
-#[derive(Debug, Clone)]
-#[cfg_attr(test, derive(Eq, PartialEq))]
-pub(crate) enum ErrorInner {
-	InvalidEscapeSequence {
-		position: usize,
-		character: Option<u8>,
-	},
-	MissingVariableName {
-		position: usize,
-		len: usize,
-	},
-	UnexpectedCharacter {
-		position: usize,
-		character: u8,
-		expected: &'static str,
-	},
-	MissingClosingBrace {
-		position: usize,
-	},
-	NoSuchVariable {
-		position: usize,
-		name: String,
-	},
-}
-
-impl From<ErrorInner> for Error {
-	fn from(inner: ErrorInner) -> Self {
-		Self { inner }
+impl From<InvalidEscapeSequence> for Error {
+	#[inline]
+	fn from(other: InvalidEscapeSequence) -> Self {
+		Self::InvalidEscapeSequence(other)
+	}
+}
+
+impl From<MissingVariableName> for Error {
+	#[inline]
+	fn from(other: MissingVariableName) -> Self {
+		Self::MissingVariableName(other)
+	}
+}
+
+impl From<UnexpectedCharacter> for Error {
+	#[inline]
+	fn from(other: UnexpectedCharacter) -> Self {
+		Self::UnexpectedCharacter(other)
+	}
+}
+
+impl From<MissingClosingBrace> for Error {
+	#[inline]
+	fn from(other: MissingClosingBrace) -> Self {
+		Self::MissingClosingBrace(other)
+	}
+}
+
+impl From<NoSuchVariable> for Error {
+	#[inline]
+	fn from(other: NoSuchVariable) -> Self {
+		Self::NoSuchVariable(other)
 	}
 }
 
 impl std::error::Error for Error {}
 
 impl std::fmt::Display for Error {
+	#[inline]
 	fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
-		match &self.inner {
-			ErrorInner::InvalidEscapeSequence { position: _, character } => {
-				if let Some(c) = character {
-					write!(f, "Invalid escape sequence: \\{}", char::from(*c))
-				} else {
-					write!(f, "Invalid escape sequence: missing escape character")
+		match self {
+			Self::InvalidEscapeSequence(e) => e.fmt(f),
+			Self::MissingVariableName(e) => e.fmt(f),
+			Self::UnexpectedCharacter(e) => e.fmt(f),
+			Self::MissingClosingBrace(e) => e.fmt(f),
+			Self::NoSuchVariable(e) => e.fmt(f),
+		}
+	}
+}
+
+/// A character or byte from the input.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum CharOrByte {
+	/// A unicode character.
+	Char(char),
+
+	/// A byte value.
+	Byte(u8),
+}
+
+impl CharOrByte {
+	/// Get the byte length of the character in the source.
+	///
+	/// For [`Self::Char`], this returns the UTF-8 lengths of the character.
+	/// For [`Self::Byte`], this is always returns 1.
+	pub fn source_len(&self) -> usize {
+		match self {
+			Self::Char(c) => c.len_utf8(),
+			Self::Byte(_) => 1,
+		}
+	}
+
+	/// Return a printable version of `self` for error messages.
+	///
+	/// The returned value implements `Display` and will print [`Self::Char`] as a quoted character,
+	/// possibly with a rust-style escape sequence.
+	///
+	/// [`Self::Byte`] is printed as zero-padded hexadecimal value, for example as `0x8E`.
```

Comment:
Bit of a nitpick: this is technically incorrect in the `is_ascii()` case... but does it really matter here? Or should we emulate [`Debug`](https://doc.rust-lang.org/std/fmt/trait.Debug.html#stability) and say that the format isn't guaranteed?

Reply:
Changed the wording (and the implementation)